File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bfs/Problem2178.java
package com.keunwon.algorithm.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem2178 {

    static int N, M;
    static int[][] map;
    static boolean[][] visited;

    final static int[][] moves = { {-1, 0}, {1, 0}, {0, 1}, {0, -1} };

    public static void main(String[] args) throws IOException {
        input();

        bfs(0, 0);
        System.out.println(map[N -1][M - 1]);
    }

    static void bfs(int r, int c) {
        Queue<Position> queue = new LinkedList<>();
        queue.add(new Position(r, c));

        visited[r][c] = true;

        while (!queue.isEmpty()) {
            Position p = queue.poll();

            for (int[] move : moves) {
                int nr = p.r + move[0];
                int nc = p.c + move[1];

                if (validMap(nr, nc) && isGo(nr, nc)) {
                    queue.add(new Position(nr, nc));
                    map[nr][nc] = map[p.r][p.c] + 1;
                    visited[nr][nc] = true;
                }
            }
        }
    }

    static boolean isGo(int r, int c) {
        return map[r][c] != 0 && !visited[r][c];
    }

    static boolean validMap(int r, int c) {
        return r >= 0 && r < N && c >= 0 && c < M;
    }

    static void input() throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        map = new int[N][M];
        visited = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            char[] input = br.readLine().toCharArray();

            for (int j = 0; j < M; j++) {
                map[i][j] = input[j] - '0';
            }
        }
    }

    static class Position {
        int r;
        int c;

        public Position(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem2178.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 미로 탐색
 * Level: 실버-1
 **/
class Problem2178 {
    private val moves = arrayOf(-1 to 0, 0 to 1, 1 to 0, 0 to -1)

    fun solution(map: Array<IntArray>): Int {
        val queue = LinkedList<Node>()
        val visited = Array(map.size) { BooleanArray(map[0].size) }

        queue.offer(Node(0, 0, 1))
        visited[0][0] = true

        while (queue.isNotEmpty()) {
            val cur = queue.poll()

            if (cur.finish(map)) return cur.distance

            for ((mr, mc) in moves) {
                val nr = cur.r + mr
                val nc = cur.c + mc

                if (nr !in map.indices || nc !in map[0].indices) continue
                if (map[nr][nc] == 0 || visited[nr][nc]) continue

                visited[nr][nc] = true
                queue.offer(Node(nr, nc, cur.distance + 1))
            }
        }
        return 0
    }

    private data class Node(
        val r: Int,
        val c: Int,
        val distance: Int,
    ) {
        fun finish(map: Array<IntArray>): Boolean {
            return r == map.lastIndex && c == map[0].lastIndex
        }
    }
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val (n, m) = br.readLine().split(" ").map { it.toInt() }
        val map = Array(n) {
            br.readLine().toCharArray().map { it.digitToInt() }.toIntArray()
        }

        Problem2178().solution(map).also { bw.write("$it") }
        bw.flush()
    }
}
